sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala

```diff
@@ -447,8 +450,18 @@ case class FlatMapGroupsWithStateExec(
         hasTimedOut,
         watermarkPresent)
 
-      // Call function, get the returned objects and convert them to rows
-      val mappedIterator = func(keyObj, valueObjIter, groupState).map { obj =>
+      // Call function
+      val iterator = try {
+        func(keyObj, valueObjIter, groupState)
+      } catch {
+        case NonFatal(e) if !e.isInstanceOf[SparkThrowable] =>
+          throw FlatMapGroupsWithStateUserFuncException(e)
+        case f: Throwable =>
+          throw f
+      }
+
+      // Get the returned objects and convert them to rows
+      val mappedIterator = iterator.map { obj =>
```

Comment:
What if the user error occurs in the iterator?  That is, if a user returns lazy iterator with custom code.